Add support for Epiphany web browser

mcatanzaro opened the pull request:
It's similar to the existing WebKit support, but a bit easier to use as
you can simply ./wpt run epiphany

Opportunistically improve a Servo error message, since we are adding
similar error messages below and consistency is nice.

Next step is to figure out how to run the tests under Weston. Currently they run in my normal desktop session, stealing input and guaranteeing that the automation tests all take forever and fail when I manually grab input back to whatever I was doing. That's why I don't have test results yet.

mcatanzaro committed:
Add support for Epiphany web browser

It's similar to the existing WebKit support, but a bit easier to use as
you can simply ./wpt run epiphany

Opportunistically improve a Servo error message, since we are adding
similar error messages below and consistency is nice.

mcatanzaro commented:
The remaining CI failure at least seems to be unrelated:

```
manifest/tests/test_manifest.py:70: in <module>
    unique_by=lambda x: x.rel_path))
E   TypeError: lists() got an unexpected keyword argument 'average_size'
```

I found the documentation of the python-hypothesis library, [and the `average_size` parameter indeed doesn't exist](https://hypothesis.readthedocs.io/en/latest/data.html?highlight=lists#hypothesis.strategies.lists). The broken code hasn't changed since https://github.com/web-platform-tests/wpt/commit/8777171a5568e9fa967ef7a807dc400d89a07637 so not sure what could have changed.

jgraham commented on tools/wptrunner/requirements_epiphany.txt:
```diff
@@ -0,0 +1,2 @@
+mozprocess == 0.26
+selenium == 3.9.0
```
I'm not sure that selenium is required since `ExecutorWebDriver` doesn't use it (some existing files are wrong here). We could probably add `mozprocess` to `requirements.txt` and eliminate a bunch of `requirements_*` files.

mcatanzaro replied:
OK, I was confused by that too, but just copy/pasted from requirements_webkit.txt.

(What is `mozprocess` for?)

jgraham commented on tools/wpt/run.py:
```diff
@@ -422,7 +422,7 @@ def setup_kwargs(self, kwargs):
 
             if binary is None:
-                raise WptrunError("Unable to find servo binary on the PATH")
+                raise WptrunError("Unable to find servo binary in PATH")
```
Not sure that this was a change important enough to make here.

mcatanzaro replied:
Er, I guess I'll leave it then? Of course it's very easy to remove if desired.

(Trivia: it most likely comes from Spanish, which uses "en" for both "in" and "on," making it rather difficult to pick the right one when writing in English.)

jgraham reviewed (approved):
This looks good, but there are a couple of changes I would appreciate before landing.

foolip commented on tools/wpt/browser.py:
```diff
@@ -758,3 +758,33 @@ def install_webdriver(self, dest=None, channel=None):
+        if output:
+            # Stable release output looks like: "Web 3.30.2"
+            # Tech Preview output looks like "Web 3.31.3-88-g97db4f40f"
+            # So split off the first word, then truncate after the first hyphen.
```
Why not include the whole 3.31.3-88-g97db4f40f? Code that processes the version string (in wpt.fyi) could make it prettier if needed, but having the full string seems potentially useful if there can be updates which don't change the leading part. @lukebjerring opinions?

lukebjerring replied:
Can keep it + adapt to this format - fyi @Hexcles - this may need to be handled in our version-string special cases.

Hexcles replied:
Is "Web" the name of the application? If so, the convention is to strip it, but all the remaining stuff can be kept, e.g. `3.30.2` or `3.31.3-88-g97db4f40f`.

mcatanzaro replied:
The codename of the project is "Epiphany" and that's probably what we'd want to display on wpt.fyi, but the actual user-visible name is just "Web". (We can also write "GNOME Web" when needed to disambiguate.)

foolip commented on tools/wptrunner/wptrunner/browsers/epiphany.py:
```diff
@@ -0,0 +1,75 @@
+from .base import get_timeout_multiplier   # noqa: F401
+from .webkit import WebKitBrowser
+from ..executors import executor_kwargs as base_executor_kwargs
+from ..executors.executorwebdriver import (WebDriverTestharnessExecutor,  # noqa: F401
```
Nice to see ExecutorWebDriver being reused :D

mcatanzaro commented:
OK, addressed all comments here. I think the only remaining problem is:

> The remaining CI failure at least seems to be unrelated:
> 
> ```
> manifest/tests/test_manifest.py:70: in <module>
>     unique_by=lambda x: x.rel_path))
> E   TypeError: lists() got an unexpected keyword argument 'average_size'
> ```
> I found the documentation of the python-hypothesis library, [and the `average_size` parameter indeed doesn't exist](https://hypothesis.readthedocs.io/en/latest/data.html?highlight=lists#hypothesis.strategies.lists). The broken code hasn't changed since [8777171](https://github.com/web-platform-tests/wpt/commit/8777171a5568e9fa967ef7a807dc400d89a07637) so not sure what could have changed.

which has me stumped, since other pull requests don't seem to be hitting this failure.

foolip commented:
@mcatanzaro, some tests only run based on what has changed in a PR, so it's possible it's already broken but just not showing up. I'll send a PR to probe for currently failing tests.

foolip commented:
Yep, the failure was already present: https://travis-ci.com/web-platform-tests/wpt/jobs/170769863

@jgraham I think we should admin merge this, but do you have final thoughts?

jgraham commented:
There are flake8 errors specific to this PR. Please fix those before merging.

foolip commented:
https://github.com/web-platform-tests/wpt/pull/14890 should fix the existing failure.

foolip commented:
But more failures will remain :(

mcatanzaro commented:
> There are flake8 errors specific to this PR. Please fix those before merging.

Hi @jgraham, where can I see these failures?

foolip commented:
@mcatanzaro flak8 is run as part of "tools/ unittests (Python 2)" and "tools/ unittests (Python 3)" on Travis. I've filed https://github.com/web-platform-tests/wpt/issues/14893 about this not being ideal.

mcatanzaro commented:
I think flake8 is fixed. There are other failures on travis, but again they look unrelated.

foolip commented:
Yep, what remains is because of https://github.com/web-platform-tests/wpt/issues/14179 and https://github.com/web-platform-tests/wpt/issues/14894, so I'll merge this now.
